### PR TITLE
White balancing

### PR DESCRIPTION
**Pull request by Mati-AAU**

### CODE REVIEW

File: app/src/main/java/dk/scuffed/whiteboardapp/pipeline/stage_combinations/FullPipeline.kt [hunk 1]
```diff
@@ -69,6 +69,13 @@ internal fun fullPipeline(
         pipeline
     )
 
+    val whitebalance = whiteBalance(
+        context,
+        perspectiveCorrection,
+        5,
+        pipeline
+    )
```
Comment (Casper-NS) [inline app/src/main/java/dk/scuffed/whiteboardapp/pipeline/stage_combinations/FullPipeline.kt:77]: Figure out how it binarization plays together with whitebalancing before merge

File: app/src/main/java/dk/scuffed/whiteboardapp/pipeline/stages/opengl_process_stages/WhitebalanceStage.kt [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+import dk.scuffed.whiteboardapp.pipeline.FramebufferInfo
+import dk.scuffed.whiteboardapp.pipeline.IPipeline
+import dk.scuffed.whiteboardapp.pipeline.stages.GLOutputStage
+
+/**
+ * Performs whitebalancing on input1 based on input2, which should represent the background color of the whiteboard.
+ */
+internal class WhitebalanceStage(
+    context: Context,
+    private val inputFramebufferInfo1: FramebufferInfo,
```
Comment (Casper-NS) [inline app/src/main/java/dk/scuffed/whiteboardapp/pipeline/stages/opengl_process_stages/WhitebalanceStage.kt:16]: Potentential for better naming of framebuffers

File: app/src/main/res/raw/saturate_shader.glsl [hunk 3]
```diff
@@ -32,8 +32,20 @@ void main() {
...
+    TURNS BLACKS FULLY BLACK
+    float isblack = smoothstep(0.05, 0.15, hsv.y);
+
+    float iswhite = step(hsv.z, 0.96);
+
+
+    if (iswhite > 0.5){
+        hsv.z *= isblack;
+    }
+*/
```
Comment (Casper-NS) [inline app/src/main/res/raw/saturate_shader.glsl:46]: We dont like commented code, either add or remove it

File: app/src/main/res/raw/whitebalance_shader.glsl [hunk 4]
```diff
@@ -0,0 +1,15 @@
+#extension GL_OES_EGL_image_external : require
+precision mediump float;
+
+uniform vec2 resolution;
+uniform sampler2D framebuffer1;
```
Comment (Casper-NS) [inline app/src/main/res/raw/whitebalance_shader.glsl:5]: Rename to better explain what is what in terms of actual picture and whitebalancing

**Review by Casper-NS — CHANGES_REQUESTED** [review]
Minor changes needed but otherwise lgtm.

File: app/src/main/java/dk/scuffed/whiteboardapp/pipeline/stage_combinations/FullPipeline.kt (hunk 1 above)
Comment (Casper-NS) [reply app/src/main/java/dk/scuffed/whiteboardapp/pipeline/stage_combinations/FullPipeline.kt:77]: works

File: app/src/main/java/dk/scuffed/whiteboardapp/pipeline/stages/opengl_process_stages/WhitebalanceStage.kt (hunk 2 above)
Comment (Casper-NS) [reply app/src/main/java/dk/scuffed/whiteboardapp/pipeline/stages/opengl_process_stages/WhitebalanceStage.kt:16]: updated

File: app/src/main/res/raw/saturate_shader.glsl (hunk 3 above)
Comment (Casper-NS) [reply app/src/main/res/raw/saturate_shader.glsl:46]: removed

**Review by Casper-NS — APPROVED** [review]

**Review by S3JER — APPROVED** [review]

**Pull request merged**